dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -213,150 +209,117 @@ StaticLinkageCheckReport findLinkageErrors() throws ClassNotFoundException, IOEx
-
-    // Breadth-first search
-    Queue<FullyQualifiedMethodSignature> queue = new ArrayDeque<>(initialMethodReferences);
-    while (!queue.isEmpty()) {
-      FullyQualifiedMethodSignature methodReference = queue.remove();
-      String className = methodReference.getClassName();
-      if (isBuiltInClassName(className)) {
-        // Ignore references to JDK package
-        continue;
+  private boolean validateMethodReferenceByClassLoader(MethodSymbolReference methodReference) {
+    String className = methodReference.getTargetClassName();
+    String methodName = methodReference.getMethodName();
+    try {
+      Class[] parameterTypes = classDumper.methodDescriptorToClass(methodReference.getDescriptor());
+      Class clazz = className.startsWith("[") ? Array.class : classDumper.loadClass(className);
+      if ("<init>".equals(methodName)) {
+        clazz.getConstructor(parameterTypes);
+      } else if ("clone".equals(methodName) && clazz == Array.class) {
+        // Array's clone method is not returned by getMethod
+        // https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getMethod-java.lang.String-java.lang.Class...-
+        return true;
+      } else {
+        clazz.getMethod(methodName, parameterTypes);
       }
+      return true;
+    } catch (NoSuchMethodException | ClassNotFoundException ex) {
+      return false;
+    }
+  }
 
-      // Case 1: we know that the class doesn't exist in the jar files
-      if (classesNotFound.contains(className)) {
-        unresolvedMethods.add(methodReference);
-        continue;
-      }
-      // Case 2: we know that the class and method exist in the jar files
-      if (availableMethodsInJars.contains(methodReference)) {
-        continue;
-      }
-
-      // Case 3: we need to check the availability of the method through the class loader
-      try {
-        if (classDumper.methodDefinitionExists(methodReference)) {
-          availableMethodsInJars.add(methodReference);
-
-          // Enqueue references from the class unless it is already visited in class usage graph
-          if (classesVisited.add(className)) {
-            ImmutableSet<FullyQualifiedMethodSignature> nextExternalMethodReferences =
-                classDumper.listExternalMethodReferences(className);
-            queue.addAll(nextExternalMethodReferences);
+  /**
+   * Returns true if the method reference has a valid referent in the classpath via BCEL API.
+   */
+  private boolean validateMethodReferenceByBcelRepository(MethodSymbolReference methodReference) {
+    String className = methodReference.getTargetClassName();
+    String methodName = methodReference.getMethodName();
+    try {
+      JavaClass javaClass = classDumper.loadJavaClass(className);
+      while (javaClass != null) {
+        Method[] methods = javaClass.getMethods();
+        // TODO: Efficient method lookup than iterating arrays for every method reference
```

Comment:
“premature optimization is the root of all evil in programming.” (Knuth, Donald. Structured Programming with go to Statements, ACM Journal Computing Surveys, Vol 6, No. 4, Dec. 1974. p.268.)


That is, delete the TODO